builder/dockerfile/evaluator.go

```diff
@@ -34,17 +36,18 @@ import (
 )
 
 // Environment variable interpolation will happen on these statements only.
-var replaceEnvAllowed = map[string]bool{
-	command.Env:        true,
-	command.Label:      true,
-	command.Add:        true,
-	command.Copy:       true,
-	command.Workdir:    true,
-	command.Expose:     true,
-	command.Volume:     true,
-	command.User:       true,
-	command.StopSignal: true,
-	command.Arg:        true,
+var replaceEnvAllowed = map[reflect.Type]bool{
+	reflect.TypeOf((*command.EnvCommand)(nil)):        true,
```

Comment:
Instead, remove this and define `ReplaceEnv` method on the commands that allow it and check for the interface in the dispatch method. If the method exists call it before calling dispatch.